src/hotspot/share/gc/shared/space.cpp

```diff
@@ -49,48 +49,42 @@
 
 HeapWord* DirtyCardToOopClosure::get_actual_top(HeapWord* top,
                                                 HeapWord* top_obj) {
-  if (top_obj != NULL) {
-    if (_sp->block_is_obj(top_obj)) {
-      if (cast_to_oop(top_obj)->is_objArray() || cast_to_oop(top_obj)->is_typeArray()) {
-        // An arrayOop is starting on the dirty card - since we do exact
-        // store checks for objArrays we are done.
-      } else {
-        // Otherwise, it is possible that the object starting on the dirty
-        // card spans the entire card, and that the store happened on a
-        // later card.  Figure out where the object ends.
-        top = top_obj + cast_to_oop(top_obj)->size();
-      }
+  if (top_obj != NULL && top_obj < (_sp->toContiguousSpace())->top()) {
+    if (cast_to_oop(top_obj)->is_objArray() || cast_to_oop(top_obj)->is_typeArray()) {
+      // An arrayOop is starting on the dirty card - since we do exact
+      // store checks for objArrays we are done.
     } else {
-      top = top_obj;
+      // Otherwise, it is possible that the object starting on the dirty
+      // card spans the entire card, and that the store happened on a
+      // later card.  Figure out where the object ends.
+      assert(_sp->block_size(top_obj) == cast_to_oop(top_obj)->size(),
+        "Block size and object size mismatch");
```

Comment:
Pre-existing: Indentation for the assert message is off.

```suggestion
      assert(_sp->block_size(top_obj) == cast_to_oop(top_obj)->size(),
             "Block size and object size mismatch");
```